python/mujinplanningclient/zmqsubscriber.py

```diff
@@ -0,0 +1,227 @@
+# -*- coding: utf-8 -*-
+# Copyright (C) 2012-2023 MUJIN Inc
+
+import threading
+
+from . import _
+from . import zmq
+from . import GetMonotonicTime
+from . import TimeoutError, UserInterrupt
+
+import logging
+log = logging.getLogger(__name__)
+
+
+class ZmqSubscriber(object):
+    """Subscriber that can handle ongoing subscriptions and automatic socket recreation.
+    """
+
+    _ctx = None # zmq context
+    _ctxown = None # created zmq context
+
+    _endpoint = None # zmq subscription endpoint
+    _endpointFn = None # function that returns the endpoint string to subscribe to, must be thread-safe
+    _callbackFn = None # function to call back when new message is received on the subscription socket
+
+    _socket = None # zmq socket
+    _lastReceivedTimestamp = 0 # when message was last received on this subscription
+    _timeout = 4.0 # beyond this number of seconds, the socket is considered dead and should be recreated
+    _checkpreemptfn = None # function for checking for preemptions
+
+    def __init__(self, endpoint, callbackFn=None, timeout=4.0, ctx=None, checkpreemptfn=None):
+        """Subscribe to zmq endpoint.
+
+        Args:
+            endpoint: the zmq endpoint string to subscribe to, or a thread-safe function that returns such string
+            callbackFn: the function to call when subscription receives latest message, it is up to the caller to decode the raw zmq message received, the argument for the callback is the received raw message
+            timeout: number of seconds, after this duration, the subscription socket is considered dead and will be recreated automatically to handle network changes (Default: 4.0)
+            checkpreemptfn: The function for checking for preemptions
+        """
+        self._timeout = timeout
+        if callable(endpoint):
+            self._endpointFn = endpoint
+        else:
+            self._endpoint = endpoint
+        self._callbackFn = callbackFn
+
+        self._ctx = ctx
+        if self._ctx is None:
+            self._ctxown = zmq.Context()
+            self._ctx = self._ctxown
+
+    def __del__(self):
+        self.Destroy()
+
+    def __repr__(self):
+        return '<%s(endpoint=%r)>' % (self.__class__.__name__, self.endpoint)
+
+    @property
+    def endpoint(self):
+        """Endpoint string for the subscription
+        """
+        return self._endpoint
+
+    def Destroy(self):
+        self._CloseSocket()
+        if self._ctxown is not None:
+            try:
+                self._ctxown.destroy()
+            except Exception as e:
+                log.exception('caught exception when destroying zmq context: %s', e)
+            self._ctxown = None
+        self._ctx = None
+
+    def _HandleReceivedMessage(self, message):
+        if self._callbackFn:
+            self._callbackFn(message)
+
+    def _HandleTimeout(self, elapsedTime):
+        if self._callbackFn:
+            self._callbackFn(None)
```

Comment:
please add a comment that says this is for resetting the user side data struct so they don't use stale data